File: emergency-room/src/zig/capture.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// capture.zig — Safe diagnostic capture modules.
//
// Best-effort, non-destructive data collection.
// PII redaction applied to all captured output (HIGH-004).
// Atomic file writes to prevent corruption (HIGH-006).
//
// Corresponds to: src/capture.v

const std = @import("std");
const utils = @import("utils");

// ── PII redaction ─────────────────────────────────────────────────────────────

const SENSITIVE_KEYS = [_][]const u8{
    "password", "passwd", "pwd",      "secret",      "token",
    "api_key",  "api-key", "apikey",
    "auth_token", "auth-token", "authtoken",
    "access_token", "access-token", "accesstoken",
    "private_key",  "private-key",  "privatekey",
    "aws_secret", "aws-secret",
    "bearer",
};

const SENSITIVE_PREFIXES = [_][]const u8{
    "akia", "abia", "acca", "asia",          // AWS keys
    "ghp_", "gho_", "ghu_", "ghs_", "ghr_", // GitHub tokens
};

/// Redact PII from `content`.  Caller owns returned slice.
pub fn redactPii(allocator: std.mem.Allocator, content: []const u8) ![]u8 {
    var lines = std.ArrayList([]u8){};
    defer {
        for (lines.items) |l| allocator.free(l);
        lines.deinit(allocator);
    }

    var iter = std.mem.splitScalar(u8, content, '\n');
    while (iter.next()) |line| {
        const redacted = try redactLine(allocator, line);
        try lines.append(allocator, redacted);
    }

    return std.mem.join(allocator, "\n", lines.items);
}

fn redactLine(allocator: std.mem.Allocator, line: []const u8) ![]u8 {
    const lower = try std.ascii.allocLowerString(allocator, line);
    defer allocator.free(lower);

    var current = try allocator.dupe(u8, line);

    // Sensitive key=value / key: value patterns.
    for (SENSITIVE_KEYS) |key| {
        if (std.mem.indexOf(u8, lower, key) != null) {
            var tmp = try redactAfterKey(allocator, current, key, '=');
            allocator.free(current);
            current = tmp;
            tmp = try redactAfterKey(allocator, current, key, ':');
            allocator.free(current);
            current = tmp;
        }
    }

    // Token-prefix patterns.
    const lower2 = try std.ascii.allocLowerString(allocator, current);
    defer allocator.free(lower2);
    for (SENSITIVE_PREFIXES) |prefix| {
        if (std.mem.indexOf(u8, lower2, prefix) != null) {
            const tmp = try redactTokenPrefix(allocator, current, prefix);
            allocator.free(current);
            current = tmp;
        }
    }

    // Private key block marker.
    const lower3 = try std.ascii.allocLowerString(allocator, current);
    defer allocator.free(lower3);
    if (std.mem.indexOf(u8, lower3, "-----begin") != null and
        std.mem.indexOf(u8, lower3, "private key") != null)
    {
        allocator.free(current);
        current = try allocator.dupe(u8, "[REDACTED PRIVATE KEY BLOCK]");
    }

    const tmp_ssn = try redactSsn(allocator, current);
    allocator.free(current);
    current = tmp_ssn;

    const tmp_email = try redactEmails(allocator, current);
    allocator.free(current);
    current = tmp_email;

    return current;
}

/// Redact the value that follows `key` + `sep` on the same line.
fn redactAfterKey(
    allocator: std.mem.Allocator,
    line: []const u8,
    key: []const u8,
    sep: u8,
) ![]u8 {
    const lower = try std.ascii.allocLowerString(allocator, line);
    defer allocator.free(lower);

    const key_pos = std.mem.indexOf(u8, lower, key) orelse return allocator.dupe(u8, line);
    const rest = line[key_pos + key.len ..];
    const sep_pos = std.mem.indexOfScalar(u8, rest, sep) orelse return allocator.dupe(u8, line);

    var value_start = key_pos + key.len + sep_pos + 1;
    if (value_start >= line.len) return allocator.dupe(u8, line);

    // Skip whitespace.
    while (value_start < line.len and (line[value_start] == ' ' or line[value_start] == '\t')) {
        value_start += 1;
    }

    // Find end of value token.
    var value_end = value_start;
    while (value_end < line.len) {
        const c = line[value_end];
        if (c == ' ' or c == '\t' or c == '\n' or c == '\r') break;
        value_end += 1;
    }

    if (value_end <= value_start) return allocator.dupe(u8, line);

    return std.fmt.allocPrint(allocator, "{s}[REDACTED]{s}", .{
        line[0..value_start],
        line[value_end..],
    });
}

/// Redact tokens that start with a known sensitive prefix.
fn redactTokenPrefix(
    allocator: std.mem.Allocator,
    line: []const u8,
    prefix: []const u8,
) ![]u8 {
    const lower = try std.ascii.allocLowerString(allocator, line);
    defer allocator.free(lower);

    const pos = std.mem.indexOf(u8, lower, prefix) orelse return allocator.dupe(u8, line);

    var end = pos + prefix.len;
    while (end < line.len) {
        const c = line[end];
        const is_tok = std.ascii.isAlphanumeric(c) or c == '_' or c == '-';
        if (!is_tok) break;
        end += 1;
    }

    if (end <= pos + prefix.len) return allocator.dupe(u8, line);

    return std.fmt.allocPrint(allocator, "{s}[REDACTED]{s}", .{
        line[0..pos],
        line[end..],
    });
}

/// Redact SSN patterns (###-##-####).
fn redactSsn(allocator: std.mem.Allocator, line: []const u8) ![]u8 {
    if (line.len < 11) return allocator.dupe(u8, line);

    var result = try allocator.dupe(u8, line);
    var i: usize = 0;

    while (i + 10 < result.len) {
        if (std.ascii.isDigit(result[i]) and
            std.ascii.isDigit(result[i + 1]) and
            std.ascii.isDigit(result[i + 2]) and
            result[i + 3] == '-' and
            std.ascii.isDigit(result[i + 4]) and
            std.ascii.isDigit(result[i + 5]) and
            result[i + 6] == '-' and
            std.ascii.isDigit(result[i + 7]) and
            std.ascii.isDigit(result[i + 8]) and
            std.ascii.isDigit(result[i + 9]) and
            std.ascii.isDigit(result[i + 10]))
        {
            const new = try std.fmt.allocPrint(allocator, "{s}[REDACTED-SSN]{s}", .{
                result[0..i],
                result[i + 11 ..],
            });
            allocator.free(result);
            result = new;
            i += 14; // len("[REDACTED-SSN]")
        } else {
            i += 1;
        }
    }

    return result;
}

fn isEmailChar(c: u8) bool {
    return std.ascii.isAlphanumeric(c) or c == '.' or c == '_' or c == '-' or c == '+' or c == '%';
}

/// Redact email addresses (word@word.word).
fn redactEmails(allocator: std.mem.Allocator, line: []const u8) ![]u8 {
    const at_pos = std.mem.indexOfScalar(u8, line, '@') orelse return allocator.dupe(u8, line);

    var start = at_pos;
    while (start > 0 and isEmailChar(line[start - 1])) {
        start -= 1;
    }

    var end = at_pos + 1;
    var has_dot = false;
    while (end < line.len) {
        const c = line[end];
        if (c == '.') {
            has_dot = true;
            end += 1;
        } else if (isEmailChar(c)) {
            end += 1;
        } else {
            break;
        }
    }

    if (!has_dot or end <= at_pos + 3 or start >= at_pos) {
        return allocator.dupe(u8, line);
    }

    return std.fmt.allocPrint(allocator, "{s}[REDACTED-EMAIL]{s}", .{
        line[0..start],
        line[end..],
    });
}

// ── Capture module types ──────────────────────────────────────────────────────

pub const CaptureResult = struct {
    name: []const u8,
    success: bool,
    output: []const u8,
    error_msg: []const u8,
    duration_ms: i64,
};

pub const CaptureModule = struct {
    name: []const u8,
    display_name: []const u8,
    commands: []const []const u8,
};

// ── Platform command lists ────────────────────────────────────────────────────

const IS_LINUX = @import("builtin").target.os.tag == .linux;
const IS_MACOS = @import("builtin").target.os.tag == .macos;
const IS_WINDOWS = @import("builtin").target.os.tag == .windows;

pub fn getOsVersionCommands() []const []const u8 {
    if (IS_LINUX) return &.{
        "cat /etc/os-release",
        "uname -a",
        "hostnamectl 2>/dev/null || true",
    };
    if (IS_MACOS) return &.{ "sw_vers", "uname -a" };
    if (IS_WINDOWS) return &.{
        "systeminfo | findstr /B /C:\"OS\"",
        "ver",
    };
    return &.{"uname -a"};
}

pub fn getUptimeCommands() []const []const u8 {
    if (IS_LINUX) return &.{ "uptime", "cat /proc/uptime" };
    if (IS_MACOS) return &.{"uptime"};
    if (IS_WINDOWS) return &.{"net statistics workstation | find \"Statistics\""};
    return &.{"uptime"};
}

pub fn getDiskCommands() []const []const u8 {
    if (IS_LINUX) return &.{
        "df -h",
        "df -i",
        "lsblk -o NAME,SIZE,FSTYPE,MOUNTPOINT 2>/dev/null || true",
    };
    if (IS_MACOS) return &.{ "df -h", "diskutil list" };
    if (IS_WINDOWS) return &.{"wmic logicaldisk get size,freespace,caption"};
    return &.{"df -h"};
}

pub fn getMemoryCommands() []const []const u8 {
    if (IS_LINUX) return &.{ "free -h", "cat /proc/meminfo | head -20" };
    if (IS_MACOS) return &.{ "vm_stat", "top -l 1 | head -10" };
    if (IS_WINDOWS) return &.{"systeminfo | findstr Memory"};
    return &.{};
}

pub fn getNetworkCommands() []const []const u8 {
    if (IS_LINUX) return &.{
        "ip addr show 2>/dev/null || ifconfig",
        "ip route show 2>/dev/null || route -n",
        "ss -tuln 2>/dev/null || netstat -tuln",
    };
    if (IS_MACOS) return &.{ "ifconfig", "netstat -rn", "netstat -an | head -50" };
    if (IS_WINDOWS) return &.{ "ipconfig /all", "netstat -an | findstr LISTENING" };
    return &.{};
}

pub fn getProcessCommands() []const []const u8 {
    if (IS_LINUX) return &.{
        "ps aux --sort=-%mem | head -20",
        "ps aux --sort=-%cpu | head -20",
    };
    if (IS_MACOS) return &.{ "ps aux | head -20", "top -l 1 -o mem | head -20" };
    if (IS_WINDOWS) return &.{"tasklist /V | findstr /V \"N/A\""};
    return &.{"ps aux | head -20"};
}

// ── RunCaptureModule ──────────────────────────────────────────────────────────

/// Run a single capture module.  Returns owned CaptureResult.
/// All string fields inside the result are owned by `allocator`.
pub fn runCaptureModule(
    allocator: std.mem.Allocator,
    mod: CaptureModule,
    logs_path: []const u8,
    dry_run: bool,
) !CaptureResult {
    const start_ns = std.time.nanoTimestamp();

    var outputs = std.ArrayList(u8){};
    defer outputs.deinit(allocator);

    var success = false;

    for (mod.commands) |cmd| {
        if (dry_run) {
            const line = try std.fmt.allocPrint(allocator, "[DRY-RUN] Would execute: {s}\n", .{cmd});
            defer allocator.free(line);
            try outputs.appendSlice(allocator, line);
            success = true;
            continue;
        }

        const result = utils.runShell(allocator, cmd) catch |e| {
            const msg = try std.fmt.allocPrint(allocator, "[WARN] Could not run '{s}': {s}\n", .{ cmd, @errorName(e) });
            defer allocator.free(msg);
            try outputs.appendSlice(allocator, msg);
            continue;
        };
        defer allocator.free(result.stdout);
        defer allocator.free(result.stderr);

        if (result.exit_code == 0) {
            const header = try std.fmt.allocPrint(allocator, "=== {s} ===\n", .{cmd});
            defer allocator.free(header);
            try outputs.appendSlice(allocator, header);
            try outputs.appendSlice(allocator, result.stdout);
            try outputs.appendSlice(allocator, "\n");
            success = true;
        }
    }

    const duration_ms = @divTrunc(std.time.nanoTimestamp() - start_ns, std.time.ns_per_ms);

    // Apply PII redaction (HIGH-004).
    const raw = outputs.items;
    const output = try redactPii(allocator, raw);
    errdefer allocator.free(output);

    // Write log file atomically (HIGH-006).
    if (!dry_run and output.len > 0) {
        const log_file = try std.fs.path.join(allocator, &.{ logs_path, try std.fmt.allocPrint(allocator, "{s}.log", .{mod.name}) });
        defer allocator.free(log_file);

        utils.atomicWriteFile(allocator, log_file, output) catch |e| {
            utils.logError(allocator, logs_path, "capture", "Failed to write log", &.{
                .{ .key = "module", .value = mod.name },
                .{ .key = "error", .value = @errorName(e) },
            });
            return CaptureResult{
                .name = mod.name,
                .success = false,
                .output = output,
                .error_msg = try std.fmt.allocPrint(allocator, "Failed to write log: {s}", .{@errorName(e)}),
                .duration_ms = @intCast(duration_ms),
            };
        };
    }

    return CaptureResult{
        .name = mod.name,
        .success = success,
        .output = output,
        .error_msg = try allocator.dupe(u8, ""),
        .duration_ms = @intCast(duration_ms),
    };
}

File: emergency-room/rust/src/capture.rs
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// Safe diagnostic capture with PII redaction.
// Ported from capture.v — same redaction rules, same capture modules.

use crate::incident::{atomic_write, CommandLog, Config, Incident};
use chrono::Utc;
use std::process::Command;

struct CaptureModule {
    name: &'static str,
    display_name: &'static str,
    commands: &'static [&'static str],
}

// ── PII redaction ────────────────────────────────────────────────────────────

const SENSITIVE_KEYS: &[&str] = &[
    "password", "passwd", "pwd", "secret", "token",
    "api_key", "api-key", "apikey",
    "auth_token", "auth-token", "authtoken",
    "access_token", "access-token", "accesstoken",
    "private_key", "private-key", "privatekey",
    "aws_secret", "aws-secret",
    "bearer",
];

const SENSITIVE_PREFIXES: &[&str] = &[
    "akia", "abia", "acca", "asia",                 // AWS
    "ghp_", "gho_", "ghu_", "ghs_", "ghr_",         // GitHub
];

pub fn redact_pii(content: &str) -> String {
    content.lines().map(redact_line).collect::<Vec<_>>().join("\n")
}

fn redact_line(line: &str) -> String {
    let lower = line.to_lowercase();
    let mut result = line.to_string();

    // Key=value / key: value patterns
    for key in SENSITIVE_KEYS {
        if lower.contains(key) {
            result = redact_after_sep(&result, key, '=');
            result = redact_after_sep(&result, key, ':');
        }
    }

    // Token prefix patterns
    for prefix in SENSITIVE_PREFIXES {
        if lower.contains(prefix) {
            result = redact_token(&result, prefix);
        }
    }

    // Private key block
    let lower_r = result.to_lowercase();
    if lower_r.contains("-----begin") && lower_r.contains("private key") {
        return "[REDACTED PRIVATE KEY BLOCK]".to_string();
    }

    result = redact_ssn(&result);
    result = redact_email(&result);
    result
}

fn redact_after_sep(line: &str, key: &str, sep: char) -> String {
    let lower = line.to_lowercase();
    let Some(key_pos) = lower.find(key) else { return line.to_string() };
    let rest = &line[key_pos + key.len()..];
    let Some(sep_pos) = rest.find(sep) else { return line.to_string() };
    let val_start_raw = key_pos + key.len() + sep_pos + 1;
    if val_start_raw >= line.len() { return line.to_string(); }

    let after_sep = &line[val_start_raw..];
    let start_offset = after_sep.len() - after_sep.trim_start().len();
    let val_start = val_start_raw + start_offset;
    if val_start >= line.len() { return line.to_string(); }

    let val_slice = &line[val_start..];
    let end = val_slice.find(|c: char| c.is_whitespace()).unwrap_or(val_slice.len());
    if end == 0 { return line.to_string(); }

    format!("{}{}{}", &line[..val_start], "[REDACTED]", &line[val_start + end..])
}

fn redact_token(line: &str, prefix: &str) -> String {
    let lower = line.to_lowercase();
    let Some(pos) = lower.find(prefix) else { return line.to_string() };
    let after = &line[pos + prefix.len()..];
    let end = after
        .find(|c: char| !(c.is_alphanumeric() || c == '_' || c == '-'))
        .unwrap_or(after.len());
    if end == 0 { return line.to_string(); }
    format!("{}{}{}", &line[..pos], "[REDACTED]", &line[pos + prefix.len() + end..])
}

fn redact_ssn(line: &str) -> String {
    let bytes = line.as_bytes();
    let mut result = String::with_capacity(line.len());
    let mut i = 0;
    while i < bytes.len() {
        if i + 10 < bytes.len()
            && bytes[i].is_ascii_digit()
            && bytes[i + 1].is_ascii_digit()
            && bytes[i + 2].is_ascii_digit()
            && bytes[i + 3] == b'-'
            && bytes[i + 4].is_ascii_digit()
            && bytes[i + 5].is_ascii_digit()
            && bytes[i + 6] == b'-'
            && bytes[i + 7].is_ascii_digit()
            && bytes[i + 8].is_ascii_digit()
            && bytes[i + 9].is_ascii_digit()
            && bytes[i + 10].is_ascii_digit()
        {
            result.push_str("[REDACTED-SSN]");
            i += 11;
        } else {
            result.push(bytes[i] as char);
            i += 1;
        }
    }
    result
}

fn redact_email(line: &str) -> String {
    let Some(at) = line.find('@') else { return line.to_string() };
    // Find start of local part
    let local_start = line[..at]
        .rfind(|c: char| !(c.is_alphanumeric() || "._%+-".contains(c)))
        .map(|p| p + 1)
        .unwrap_or_else(|| 0);
    // Find end of domain
    let domain = &line[at + 1..];
    let domain_end = domain
        .find(|c: char| !(c.is_alphanumeric() || ".-".contains(c)))
        .unwrap_or(domain.len());
    let has_dot = domain[..domain_end].contains('.');
    if has_dot && domain_end > 2 && at > local_start {
        format!(
            "{}[REDACTED-EMAIL]{}",
            &line[..local_start],
            &line[at + 1 + domain_end..]
        )
    } else {
        line.to_string()
    }
}

// ── Capture modules ──────────────────────────────────────────────────────────

#[cfg(target_os = "linux")]
const MODULES: &[CaptureModule] = &[
    CaptureModule {
        name: "os_version",
        display_name: "OS Version",
        commands: &["cat /etc/os-release", "uname -a"],
    },
    CaptureModule {
        name: "uptime",
        display_name: "System Uptime",
        commands: &["uptime", "cat /proc/uptime"],
    },
    CaptureModule {
        name: "disk_free",
        display_name: "Disk Space",
        commands: &["df -h", "df -i"],
    },
    CaptureModule {
        name: "memory",
        display_name: "Memory Status",
        commands: &["free -h"],
    },
    CaptureModule {
        name: "network_summary",
        display_name: "Network Summary",
        commands: &["ip addr show", "ip route show"],
    },
    CaptureModule {
        name: "process_summary",
        display_name: "Process Summary",
        commands: &["ps aux --sort=-%mem", "ps aux --sort=-%cpu"],
    },
];

#[cfg(target_os = "macos")]
const MODULES: &[CaptureModule] = &[
    CaptureModule { name: "os_version",       display_name: "OS Version",       commands: &["sw_vers", "uname -a"] },
    CaptureModule { name: "uptime",           display_name: "System Uptime",    commands: &["uptime"] },
    CaptureModule { name: "disk_free",        display_name: "Disk Space",       commands: &["df -h"] },
    CaptureModule { name: "memory",           display_name: "Memory Status",    commands: &["vm_stat"] },
    CaptureModule { name: "network_summary",  display_name: "Network Summary",  commands: &["ifconfig", "netstat -rn"] },
    CaptureModule { name: "process_summary",  display_name: "Process Summary",  commands: &["ps aux"] },
];

#[cfg(target_os = "windows")]
const MODULES: &[CaptureModule] = &[
    CaptureModule { name: "os_version",       display_name: "OS Version",       commands: &["ver"] },
    CaptureModule { name: "disk_free",        display_name: "Disk Space",       commands: &["wmic logicaldisk get size,freespace,caption"] },
    CaptureModule { name: "memory",           display_name: "Memory Status",    commands: &["systeminfo | findstr Memory"] },
    CaptureModule { name: "network_summary",  display_name: "Network Summary",  commands: &["ipconfig /all"] },
    CaptureModule { name: "process_summary",  display_name: "Process Summary",  commands: &["tasklist /V"] },
];

#[cfg(not(any(target_os = "linux", target_os = "macos", target_os = "windows")))]
const MODULES: &[CaptureModule] = &[
    CaptureModule { name: "os_version", display_name: "OS Version", commands: &["uname -a"] },
];

pub fn run_all(inc: &mut Incident, cfg: &Config) {
    for module in MODULES {
        let ok = run_module(module, inc, cfg);
        if ok {
            println!("  \x1b[32m✓\x1b[0m {}", module.display_name);
        } else {
            println!("  \x1b[33m○\x1b[0m {} (skipped)", module.display_name);
        }

        let now = Utc::now().to_rfc3339();
        inc.commands.push(CommandLog {
            name: module.name.to_string(),
            command: module.commands.join(" | "),
            started_at: now.clone(),
            ended_at: now,
            exit_code: if ok { 0 } else { 1 },
            output_len: 0,
        });
    }

    // Update receipt with command log
    let _ = crate::incident::write_envelope(inc, cfg);
}

fn run_module(module: &CaptureModule, inc: &Incident, cfg: &Config) -> bool {
    let mut parts: Vec<String> = Vec::new();
    let mut any_ok = false;

    for &cmd in module.commands {
        if cfg.dry_run {
            parts.push(format!("[DRY-RUN] Would execute: {cmd}"));
            any_ok = true;
            continue;
        }

        // Split command into argv to avoid shell injection
        let argv: Vec<&str> = cmd.split_whitespace().collect();
        if argv.is_empty() { continue; }
        let output = Command::new(argv[0]).args(&argv[1..]).output();
        if let Ok(out) = output {
            if out.status.success() {
                parts.push(format!("=== {cmd} ==="));
                parts.push(String::from_utf8_lossy(&out.stdout).into_owned());
                parts.push(String::new());
                any_ok = true;
            }
        }
    }

    if !any_ok || cfg.dry_run { return any_ok; }

    let raw = parts.join("\n");
    let redacted = redact_pii(&raw);
    let log_file = inc.logs_path.join(format!("{}.log", module.name));

    if let Err(e) = atomic_write(&log_file, &redacted) {
        eprintln!("\x1b[33m[WARN]\x1b[0m capture {}: {e}", module.name);
        return false;
    }
    true
}

// ── Tests ────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn redacts_api_key() {
        let line = "API_KEY=supersecretvalue123";
        let out = redact_line(line);
        assert!(out.contains("[REDACTED]"), "expected redaction in: {out}");
        assert!(!out.contains("supersecretvalue"), "value should be gone: {out}");
    }

    #[test]
    fn redacts_github_token() {
        let line = "token: ghp_ABCDEF1234567890";
        let out = redact_line(line);
        assert!(out.contains("[REDACTED]"), "expected redaction: {out}");
    }

    #[test]
    fn redacts_ssn() {
        let line = "SSN: 123-45-6789 on file";
        let out = redact_ssn(line);
        assert!(out.contains("[REDACTED-SSN]"), "expected SSN redaction: {out}");
    }

    #[test]
    fn redacts_email() {
        let line = "contact: user@example.com please";
        let out = redact_email(line);
        assert!(out.contains("[REDACTED-EMAIL]"), "expected email redaction: {out}");
    }

    #[test]
    fn innocent_text_unchanged() {
        let line = "uptime: 3 days, 4 hours";
        assert_eq!(redact_line(line), line);
    }
}
